src/lms5xx_node.cpp

```diff
@@ -0,0 +1,263 @@
+            " on the \"multi_echo\" topic in addition to the regular, single echo \"scan\" topic." << std::endl << std::endl;
+  std::cout << "Parameters:" << std::endl;
+  std::cout << "    host      The IP of the scanner" << std::endl;
+  std::cout << "    port      The port to connect on" << std::endl;
+  std::cout << "    frame_id  Frame id of the laser, defaults to \"laser\"." << std::endl;
+  std::cout << "    echoes    One of \"first\", \"last\" or \"all\"." << std::endl;
+}
+
+bool setup(LMS5xx &laser, sensor_msgs::LaserScan &scan_msg, sensor_msgs::MultiEchoLaserScan &multi_scan_msg,
+           const LMS5xx::EchoFilter echo_mode)
+{
+  ScanConfig cfg;
+  ScanOutputRange output_range;
+  ScanDataConfig dataCfg;
+
+  ROS_DEBUG("Logging in to laser.");
+  laser.login();
+  cfg = laser.get_scan_config();
+  output_range = laser.get_scan_output_range();
+
+  if (cfg.scan_frequency != 5000)
+  {
+    laser.disconnect();
+    ROS_WARN("Unable to get laser output range. Retrying.");
+    ros::Duration(1).sleep();
+    return false;
+  }
+
+  ROS_INFO("Connected to laser.");
+
+  ROS_DEBUG("Laser configuration: scaningFrequency %d, activeSensors %d, angleResolution %d, startAngle %d, stopAngle %d",
+           cfg.scan_frequency, cfg.num_sectors, cfg.angualar_resolution, cfg.start_angle, cfg.stop_angle);
+  ROS_DEBUG("Laser output range: angleResolution %d, startAngle %d, stopAngle %d",
+           output_range.angular_resolution, output_range.start_angle, output_range.stop_angle);
+
+  scan_msg.range_min = 0.01;
+  scan_msg.range_max = 20.0; // TODO: value?
+  scan_msg.scan_time = 100.0 / cfg.scan_frequency;
+  scan_msg.angle_increment = (double)output_range.angular_resolution / 10000.0 * DEG2RAD;
+  scan_msg.angle_min = (double)output_range.start_angle / 10000.0 * DEG2RAD - M_PI / 2;
+  scan_msg.angle_max = (double)output_range.stop_angle / 10000.0 * DEG2RAD - M_PI / 2;
+
+  multi_scan_msg.range_min = 0.01;
+  multi_scan_msg.range_max = 20.0; // TODO: value?
+  multi_scan_msg.scan_time = 100.0 / cfg.scan_frequency;
+  multi_scan_msg.angle_increment = (double)output_range.angular_resolution / 10000.0 * DEG2RAD;
+  multi_scan_msg.angle_min = (double)output_range.start_angle / 10000.0 * DEG2RAD - M_PI / 2;
+  multi_scan_msg.angle_max = (double)output_range.stop_angle / 10000.0 * DEG2RAD - M_PI / 2;
+
+  ROS_DEBUG_STREAM("Device resolution is " << (double)output_range.angular_resolution / 10000.0 << " degrees.");
+  ROS_DEBUG_STREAM("Device frequency is " << (double)cfg.scan_frequency / 100.0 << " Hz");
+
+  int angle_range = output_range.stop_angle - output_range.start_angle;
+  int num_values = angle_range / output_range.angular_resolution ;
+  if (angle_range % output_range.angular_resolution == 0)
+  {
+    // Include endpoint
+    ++num_values;
+  }
+  scan_msg.ranges.resize(num_values);
+  scan_msg.intensities.resize(num_values);
+
+  multi_scan_msg.ranges.resize(5);
```

Comment:
Yes, this is always 5 as the only setting for multiple echoes is "all" and the sensor returns 5 echoes.